### DISCUSSION TITLE
what's the difference between `/usr/local/bin/` and `/usr/local/opt/` on macOS?

### DISCUSSION THREAD

**Discussion by xiebruce** [Homebrew/discussions#242] [Everyday usage]
I found that brew will create symbolic link from both `/usr/local/bin/` and `/usr/local/opt/` to `/usr/local/Cellar/xxx/xx.xx/xx` after installing a software, what's the difference between `/usr/local/bin/` and `/usr/local/opt/`? because this two directories seems to be duplicates.

![image](https://user-images.githubusercontent.com/13575500/101459014-a7d8f280-3972-11eb-8cde-2f05d3220104.png)

**Comment by gromgit** [comment]
See #239 for a related discussion, but they are definitely _not_ duplicates, and they serve different purposes.

`/usr/local/opt` contains links to the _base_ of the "active" Cellar directory for each formula, and serves as a stable reference to the formula itself. This is so that, for instance, anything that needs to link to `libpng` can pass the fixed flag `-L/usr/local/opt/libpng/lib` to the C compiler, instead of having to dynamically determine the current versioned path to that same directory (now `/usr/local/Cellar/libpng/1.6.37/lib`, previously `/usr/local/Cellar/libpng/1.6.36/lib`...or was is `/usr/local/Cellar/libpng/1.6.33/lib`?) `/usr/local/opt` links are updated on each `brew install/upgrade`, so your binary doesn't have to be recompiled when you upgrade `libpng` from 1.6.36 to 1.6.37, and the old lib directory suddenly disappears.

`/usr/local/bin` contains links to the _contents of the bin directory_ of the "active" Cellar directory for each formula, and server as a stable directory that can be added to your `PATH` to instantly gain access to every utility in every formula in one place.

The two are usually "in sync", but not always. One common counterexample: _keg-only_ formulae will be linked into `/usr/local/opt` as always, but not into `/usr/local/...`.
